### ISSUE TITLE
Set priority segments to allow vehicles to enter blocked junction

### ISSUE THREAD

**Issue by aubergine10** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#424]
When setting priority signs, it would be nice if they could affect the "Enter blocked junction" setting either automatically (via mod option) or with keybind modifier.

* Priority segment = allow enter blocked junction
* Yield / Stop segment = deny enter blocked junction

**Comment by aubergine10** [comment]
Brain: Why even have 'enter blocked junction' when we have priority signs?

Also brain: No idea.

Also brain: Why not merge priority signs in to junction restrictions, and ditch the 'enter blocked junction' thing?

**Comment by rafamariofan** [comment]
Isn't "(don't)enter blocked junctions" a more CPU lightweight option than TMPE priority signs since it's something that is in the vanilla game?

At least that's the exact reason why I preffer to just use this setting instead of priority Signs in my roundabouts lol.

**Comment by kianzarrin** [comment]
In a 4 way junction with priority signs people on the major road should not want to enter blocked junctions for people on the minor road to be able to go into directions that is not blocked.
ideally  software can automatically understand if we should enable entering blocked junctions according to common situations, though I suspect its going to be really hard considering people have all sort of custom junctions.

**Comment by brunoais** [comment]
I agree with kianzarrin.
I use priority signs and "can enter in blocked junctions" differently depending on multiple factors, including what rafamariofan and kianzarrin said together. Please don't take that away.

As for the feature that is suggested here... Yes please! I like such default.

**Comment by aubergine10** [comment]
> Isn't "(don't)enter blocked junctions" a more CPU lightweight option than TMPE priority signs since it's something that is in the vanilla game?

The feature in vanilla is the "Stop" sign (**traffic routes > junctions** info view) - and we already have that covered in TM:PE priority signs. So the "don't enter blocked junctions", as far as I know, isn't even part of vanilla.

Vanilla essentially has two signs: Yield (default) and Stop. The priority _routes_ feature in vanilla, as far as I can tell, has two purposes:

1. It automatically adds stop signs for side roads at junctions (which is often very annoying)
2. It tells snow ploughs which roads to prioritise for snow clearing (I've still never played a Snow biome map so not sure if that's correct)

> In a 4 way junction with priority signs people on the major road should not want to enter blocked junctions for people on the minor road to be able to go into directions that is not blocked.

Isn't that what TM:PE priority signs are for? The Priority/Yield/Stop sign (default: Yield) for each segment at the junction controls how traffic should enter the junction (however, see additional note below).

> ideally software can automatically understand if we should enable entering blocked junctions according to common situations, though I suspect its going to be really hard considering people have all sort of custom junctions.

Any lane that would cross oncoming traffic (eg. left or right turn depending on what side of road cars drive on) should always yield to avoid collisions. That's a "road rule" that applies everywhere: Don't crash in to other cars. So that should be baked in to the priority sign logic - it shouldn't require is to fiddle with "enter blocked junction" stuff.

For complex junctions we could use conflict point detection (see #4 for details) to determine which _lanes_ must yield at the junction.

> If I have a 3 lane road with L/S/R directions on each lane, going to a 4-way intersection. Only the turning left has to wait because too much traffic, the others could cross right away... Keeping this one can lead all lanes to be blocked because the traffic in the main road is just there waiting on the way and the other cars are blocking the whole block.

See my comments above. The yielding when crossing oncoming traffic should be baked in to priority sign logic. We shouldn't have to fudge it with "allow enter blocked junction" stuff.

Also, I think the whole implementation of "allow enter blocked junction" is wrong in TM:PE. It shouldn't be defined segment by segment (we have Priority/Yield/Stop signs for that), it should be defined in the junction node. I can't speak for other countries, but in UK we have this:

![image](https://user-images.githubusercontent.com/1386719/68734733-4d386000-05d3-11ea-830c-dde82e70cafc.png)

So on a normal junction, everything is handled by priority rules. But some super-busy junctions, which can lead to gridlock if they get blocked, have that extra "keep this junction clear" rule (the yellow hash pattern on the road).

I think all too often users are using "enter blocked junction" feature as a weird alternate to priority/yield/stop signs. For good traffic flow, generally speaking, vehicles should be able to enter junctions if they are clear (ie. default of Yield for all roads entering a junction, unless user defines otherwise).

There are some cases where you want to keep junctions clear of stagnant traffic, most notably level crossings (or busy junctions like the image above). But in those cases the "enter blocked junction" setting applies to the _junction_ itself, not the roads entering the junction.

My suggestion is therefore:

1. Improve the Priority/Yield/Stop signs to be 'conflict point'-aware, and yield intelligently for lanes that would cross other traffic
2. Retire existing "enter blocked junction"
3. Add new "Keep junction clear" feature, which applies to the _node_, in the Priority Signs (or junction restrictions?) UI.
4. Level crossings would automatically get "keep junction clear" (for freaking obvious reasons), and maybe junctions with tram routes too? User could still override on junction by junction basis.

If even more control is required, traffic lights should be used to fully control traffic across the junction.

**Comment by aubergine10** [comment]
Some other thoughts:

* Box junctions (image above) are just a hacky way of dealing with symptoms which manifest in the road network due to problems elsewhere
    * It's generally best idea to find the root problem, and deal with that at source (which in many cases could simply be providing better public transport)
    * Sometimes fixing the root problem would be too much grief (eg. requiring buildings to be destroyed or excessive road works), so it's easier to apply the "keep junction clear" hack
* There is sometimes benefit to allowing vehicles to enter blocked level crossing, in cases where it's close to train station and there are multiple tracks to cross (see also: #11)

**Comment by aubergine10** [comment]
Maybe an enum at the node could define traffic behaviour through the node:

- `NormalJunction`
    - "Try not to crash"
- `WaitInJunction`
    - for larger junctions, traffic can partially enter and wait in junction 
    - see #46 (particularly the youtube video in OP of that issue)
    - if an entering road has bike lanes, it would behave like fourth pic in #14
- `BoxJunction`
    -  traffic cannot enter until previous traffic cleared
    - replaces segment-based 'enter blocked junction' with node-based 'box junction'
- `MiniRoundabout`
    - instead of traffic driving in arc towards exit lane, it goes in cicular motion as if the node is a roundabout - see #73
- `LaneRestriction`
    - traffic calming measure (small roads only) - first pic in #14
- `Crossing`
    - Similar to Crossings mod but maybe with some refinements? - second pic in #14
- `SpeedBump`
    - Node acts like a speed bump (small roads only), nerfing traffic speed on adjacent segments - see #68

Priority signs then define who has right of way when entering junction. Traffic Lights override priority signs, but could be affected by them (eg. priority route gets more green time or more beneficial flow ratios etc).

**Comment by kianzarrin** [comment]
It is true that users get confused by enter blocked junction. I know I did. But this problem is soon diminished when we allow users to setup intent. Intent can be used for generic stuff and for more advanced users who like flexibility they can micro manage a bit more for all the crazy junctions they have. 

@aubergine10 I think this is a bad idea to take that flexibility away. We should let the user decide to enter blocked junctions or not on a segment end by segment end basis. TMPE users like flexibility. To automatically decide to enter blocked junctions or not can be complicated. If we are going to take that flexibility away from the user  then: Either do it perfect or don't do it at all. 

I am worried that this is going to be a repeat of the highway rules mistake that people don't use because they don't like extra restrictions.

I also think that the code for enter blocked junction can improve. for example consider a big highway roundabout. once I disable enter blocked junctions, cars entering the roundabout wait for the junction to be clear before entering. This happens even if traffic is flowing fast. I think if traffic is flowing fast there is no need for cars entering the roundabout to wait a little bit longer. For this reason in my highway roundabouts, I generally set enter blocked junction for yielding cars entering the roundabout. The traffic just flows faster that way. for my urban roundabouts yielding cars cannot enter blocked junctions. 

So we need to solve some problems including the one I mentioned above to abolish enter blocked junctions and I think we are no where near ready. Also I think its not worth the time specially considering that the **intent** feature is coming.

**Comment by kianzarrin** [comment]
> Also, I think the whole implementation of "allow enter blocked junction" is wrong in TM:PE. It shouldn't be defined segment by segment (we have Priority/Yield/Stop signs for that), it should be defined in the junction node. I can't speak for other countries, but in UK we have this: [box junction]

@aubergine10  this is in contradiction to another thing you said:

> Any lane that would cross oncoming traffic (eg. left or right turn depending on what side of road cars drive on) should always yield to avoid collisions. That's a "road rule" that applies everywhere: Don't crash in to other cars. So that should be baked in to the priority sign logic - it shouldn't require is to fiddle with "enter blocked junction" stuff.

So first you say enter blocked junction belongs to node where it can be modified manually. Then you say it belongs to segment end but should be automatic.
 can think of situations where some cars can enter blocked junction and some cars cannot. for example T junctions:
![Screenshot (934)](https://user-images.githubusercontent.com/26344691/80411563-0f0ad080-88d5-11ea-81de-68b80d9330b8.png)

Again making it automatic is a do it perfect or don't do it at all thing.

**Comment by aubergine10** [comment]
In the image above, it shows another issue we have - the priority signs should factor in 'does this lane cross traffic'. A similar connundrum is discussed in #565.

**Comment by aubergine10** [comment]
For anyone following this issue #424 please go take a look at the awesome stuff being worked on in #1229 (specifically the videos near bottom of that issue) - very significant improvements to both normal junctions, and also those with priority signs.

Example of normal junction with turning lane - no priority signs, etc., just junction with lane-arrow based turning lane:

https://user-images.githubusercontent.com/19638970/152700929-041caea3-6e15-493c-9af9-067901eff3e8.mp4

**Comment by brunoais** [comment]
@aubergine10 What's that mod for traffic scenario? It would be great for my junction testing as I'm experimenting with different junctions

**Comment by krzychu124** [comment]
It's my internal mod for testing. I plan to release it as a debug tool but for now... it's just never-ending sandbox where I throw my prototypes and ideas.

**Comment by brunoais** [comment]
@krzychu124 Thanks for sharing that. When you are comfortable enough, I'll make time to test it and also to test my own stuff.
